include/envoy/router/route_config_provider_manager.h

```diff
@@ -37,26 +37,26 @@ class RouteConfigProviderManager {
    * @param stat_prefix supplies the stat_prefix to use for the provider stats.
    * @param init_manager supplies the init manager.
    */
-  virtual RouteConfigProviderSharedPtr getRouteConfigProvider(
+  virtual RouteConfigProviderSharedPtr getRdsRouteConfigProvider(
       const envoy::config::filter::network::http_connection_manager::v2::Rds& rds,
       Upstream::ClusterManager& cm, Stats::Scope& scope, const std::string& stat_prefix,
       Init::Manager& init_manager) PURE;
-};
 
-/**
- * The ServerRouteConfigProviderManager additionally allows listing all of the currently managed
- * RouteConfigProviders.
- */
-class ServerRouteConfigProviderManager : public RouteConfigProviderManager {
-public:
-  virtual ~ServerRouteConfigProviderManager() {}
+  virtual RouteConfigProviderSharedPtr
+  getStaticRouteConfigProvider(envoy::api::v2::RouteConfiguration route_config,
```

Comment:
`const envoy::api::v2::RouteConfiguration&`. Also can we add some API docs for this new function?

Reply:
Will add docs.

Re: pass by const& vs value, this isn't explicitly addressed in our/G's style guide? I'm sure you already know the reasons I like it by value but I wouldn't fight about it save perf-sensitive paths, which this obviously isn't.

Reply:
We should flesh it out. In general the existing code follows roughly the following guidelines.

"For small objects, pass by value. For large objects, prefer passing by const-reference. If it's clear that move semantics will yield better performance, pass by rvalue reference with an explicit && operator. The reason for the explicit rvalue reference indication is that the call site then must either pass a temporary or use std::move() which makes the intention more clear." (The latter part is already covered when discussing unique_ptr, etc.).

Reply:
I, for one, try to always pass by value when possible (and move() to take advantage) aside from primitives. This is reflexive enough for me that I want to get ahead of whether you/other maintainers care enough about this that I have to change that.

So, to formalize my take:
1. For cheap-copy objects, e.g. primitives and some POD, pass by value
2. If the function *needs its own copy of the object for any reason*, pass by value 
3. Otherwise, pass by const reference

The additional restrictions you've laid out above, while fine for existing code, don't really make sense to me going forward. They complicate things for the caller instead of giving the caller the *option* of whether they care about perf. Passing by `const&` precludes taking advantage of move semantics, and passing by `&&` precludes anything *but* rvalues. Passing by value gives you both.
 
It's obviously possible to have the internet's millionth _long_ discussion about this but I also don't see it as something where consistency is critical, like with many extant conventions. In any event let's take it off this PR if necessary.

Reply:
> This is reflexive enough for me that I want to get ahead of whether you/other maintainers care enough about this that I have to change that.

Yes, we are care about consistency.

> It's obviously possible to have the internet's millionth long discussion about this but I also don't see it as something where consistency is critical, like with many extant conventions. In any event let's take it off this PR if necessary.

The expectation is that in general consistency is followed unless there is a good reason not to, which in this case is a matter of opinion. Happy to discuss further offline but at this point I'm going to consider this thread closed and assume that you will be consistent in this area. Thanks!